ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ClasspathComputer.java

```diff
@@ -57,86 +58,125 @@ public class ClasspathComputer {
 		JavaCore.create(project).setRawClasspath(entries, null);
 	}
 
-	public static IClasspathEntry[] getClasspath(IProject project, IPluginModelBase model, Map<String, IPath> sourceLibraryMap, boolean clear, boolean overrideCompliance) throws CoreException {
-		IJavaProject javaProject = JavaCore.create(project);
-		ArrayList<IClasspathEntry> result = new ArrayList<>();
-		IBuild build = getBuild(project);
+	public static IClasspathEntry[] getClasspath(IProject project, IPluginModelBase model,
+			Map<String, IPath> sourceLibraryMap, boolean clear, boolean overrideCompliance) throws CoreException {
+		ClasspathComputer cpc = new ClasspathComputer(project, model, sourceLibraryMap, clear, overrideCompliance);
+		return cpc.compute();
+	}
+
+	private IClasspathEntry[] compute() throws CoreException {
 
 		// add JRE and set compliance options
 		String ee = getExecutionEnvironment(model.getBundleDescription());
-		result.add(createEntryUsingPreviousEntry(javaProject, ee, PDECore.JRE_CONTAINER_PATH));
-		setComplianceOptions(JavaCore.create(project), ee, overrideCompliance);
+		addContainerEntry(getEEPath(ee));
+		setComplianceOptions(javaProject, ee, overrideCompliance);
 
 		// add pde container
-		result.add(createEntryUsingPreviousEntry(javaProject, ee, PDECore.REQUIRED_PLUGINS_CONTAINER_PATH));
+		addContainerEntry(PDECore.REQUIRED_PLUGINS_CONTAINER_PATH);
 
 		// add own libraries/source
-		addSourceAndLibraries(project, model, build, clear, sourceLibraryMap, result);
+		addSourceAndLibraries();
 
-		IClasspathEntry[] entries = result.toArray(new IClasspathEntry[result.size()]);
-		IJavaModelStatus validation = JavaConventions.validateClasspath(javaProject, entries, javaProject.getOutputLocation());
+		IClasspathEntry[] entries = collectInOriginalOrder();
+
+		IJavaModelStatus validation = JavaConventions.validateClasspath(javaProject, entries,
+				javaProject.getOutputLocation());
 		if (!validation.isOK()) {
 			PDECore.logErrorMessage(validation.getMessage());
 			throw new CoreException(validation);
 		}
+		return entries;
+	}
+
+	private IClasspathEntry[] collectInOriginalOrder() {
+		List<IClasspathEntry> result = new ArrayList<>();
+		// also collect original entries not explicitly reloaded
+		Stream<IClasspathEntry> collecting = Stream.concat(reloaded.stream(), Stream.of(original));
+		// in original order, collect unique entries only, and
+		// prevent them from being collected as new entries below
+		Map<IPath, IClasspathEntry> collectingByKey = collecting
+				.collect(Collectors.toMap(e -> keyOfPath(e.getPath()), e -> e, (first, dupe) -> first));
+		result.addAll(Stream.of(original).map(e -> collectingByKey.remove(keyOfPath(e.getPath())))
+				.filter(e -> e != null).collect(Collectors.toList()));
+		// append new entries in the order they were loaded
+		result.addAll(reloaded.stream().filter(e -> collectingByKey.remove(keyOfPath(e.getPath())) != null)
+				.collect(Collectors.toList()));
```

Comment:
```suggestion
		Arrays.stream(original).map(e -> collectingByKey.remove(keyOfPath(e.getPath())))
				.filter(e -> e != null).forEachOrdered(result::add);
		// append new entries in the order they were loaded
		reloaded.stream().filter(e -> collectingByKey.remove(keyOfPath(e.getPath())) != null)
				.forEachOrdered(result::add);
```
This avoids the creation of intermediate List-objects.

Reply:
done